app/server/endpoint/src/main/java/io/syndesis/server/endpoint/v1/handler/meta/StepMetadataHelper.java

```diff
@@ -0,0 +1,135 @@
+import io.syndesis.common.model.action.Action;
+import io.syndesis.common.model.action.ActionDescriptor;
+import io.syndesis.common.model.integration.Step;
+
+/**
+ * @author Christoph Deppisch
+ */
+final class StepMetadataHelper {
+
+    static final DataShape ANY_SHAPE = new DataShape.Builder().kind(DataShapeKinds.ANY).build();
+    static final DataShape NO_SHAPE = new DataShape.Builder().kind(DataShapeKinds.NONE).build();
+
+    /**
+     * Prevent instantiation of utility class.
+     */
+    private StepMetadataHelper() {
+        super();
+    }
+
+    /**
+     * Extracts variants from given original data shape excluding the given variant. Includes the original data shape itself as variant to the list of extracted variants.
+     * In case given original data shape and exclude variant happen to be equal just return original data shape itself as exclusive variant and set given variant meta data.
+     * @param original the original data shape providing variants
+     * @param variant variant to exclude from the original variants
+     * @param variantMeta optional new meta data variant value in case original and exclude happen to be equal
+     * @return
+     */
+    static List<DataShape> extractVariants(DataShape original, DataShape variant, String variantMeta) {
+        if (original.equals(variant)) {
+            return Collections.singletonList(new DataShape.Builder()
+                    .createFrom(original)
+                    .putMetadata(DataShapeMetaData.VARIANT, variantMeta)
+                    .variants(Collections.emptyList())
+                    .build());
+        } else {
+            List<DataShape> variants = original.getVariants()
+                    .stream()
+                    .filter(shape -> !shape.equals(variant))
+                    .collect(Collectors.toList());
+
+            variants.add(new DataShape.Builder()
+                    .createFrom(original)
+                    .variants(Collections.emptyList())
+                    .build());
+
+            return variants;
+        }
+    }
+
+    /**
+     * Evaluates input data shape on given step to be present and to not be of kind
+     * {@link DataShapeKinds#NONE}
+     * @param step
+     * @return
+     */
+    static boolean hasInputDataShape(Step step) {
+        return !getInputDataShape(step).getKind().equals(DataShapeKinds.NONE);
+    }
+
+    /**
+     * Evaluates output data shape on given step to be present and to not be of kind
+     * {@link DataShapeKinds#NONE}
+     * @param step
+     * @return
+     */
+    static boolean hasOutputDataShape(Step step) {
+        return !getOutputDataShape(step).getKind().equals(DataShapeKinds.NONE);
+    }
+
+    /**
+     * Safely get the input data shape of given step. If shape is not present this method returns
+     * a dummy data shape of kind {@link DataShapeKinds#NONE}.
+     * @param step
+     * @return
+     */
+    static DataShape getInputDataShape(Step step) {
```

Comment:
These methods could be on the `WithDataShapes` interface thinggy, and `Step` could pull them in that way, so we'd have `Step::inputDataShape()` method or something.

If you decide to go that way, make sure that you place `@Value.Auxiliary` or `@Value.Derived` and `@JsonIgnore` annotations on those methods.

Reply:
Hmm, just saw that `Step` indeed has a `inputDataShape` and that I added it, could it be used instead of this method?